### PR TITLE
ui, sql: surface Implicit Transactions for Statements

### LINKED ISSUE #36009: ui: Accurately capture statement performance for pipelined writes
When a transaction contains at least one write statement, the writes will now (as of 2.1) be pipelined: the db acknowledges a write almost as soon as the statement is issued (< 5 ms if gateway=leaseholder) while the consensus replication is carried out behind the scenes in the transaction. Any waiting on consensus happens on commit rather than synchronously when the statement is issued.

In the statements page, an unsuspecting user might see the statement performance and think their performance is incredible, when in actuality they are not seeing the larger picture of their complete transaction performance.

We should capture the wait that happens in the commit phase of the transaction (perhaps per transaction fingerprint) and report that back to the user so that they will have a more accurate picture of true performance.

cc @awoods187

### PR DESCRIPTION
**Pull request by celiala**
Fixes: #36009.

This PR surfaces whether or not a statement was run in an implicit transaction for the web ui, so it's possible to infer (for example) whether or not the statement timing includes its replication latency (See #36009).

This involves:
- sql: Adding an `implicit_txn` flag to `node_statement_statistics` and `crdb_internal.node_statement_statistics`
- ui: surfacing `implicit_txn` on the Statement Details page
- ui: grouping statements by both statement and implicit_txn on both the Statement Details and Statements (Summary) pages

Statement Summary Page:
Txns Summary

Statement Summary, Tooltip:
Txn Tooltip

Statement Details, Implicit Transaction:
Txns Implicit

Statement Details, Explicit Transaction:

Txn Explicit

Release note (admin ui change): Adding `Implicit Transaction` flag to Statements and Statement Details page.

### CODE REVIEW

File: pkg/roachpb/app_stats.proto [hunk 1]
```diff
@@ -111,6 +111,7 @@ message StatementStatisticsKey {
   optional bool distSQL = 3 [(gogoproto.nullable) = false];
   optional bool failed = 4 [(gogoproto.nullable) = false];
   optional bool opt = 5 [(gogoproto.nullable) = false];
+  optional bool implicitTransaction = 6 [(gogoproto.nullable) = false];
```
Comment (vilterp) [inline pkg/roachpb/app_stats.proto]: nit: style in protos is `implicit_transaction`

**Comment by nvanbenschoten** [comment]
@celiala thanks for writing this up. Any way I can help push it over the finish line?

**Comment by celiala** [comment]
Thanks for the ping, Nathan: 
Jordan and I just sync'ed on how to finish this up -- should have an updated PR up soon.

**Comment by celiala** [comment]
This is RFAL, thanks!

- Paired with Jordan to finish up PR #36340 & combine into this commit.
- Fixed proto naming nit.

**Comment by celiala** [comment]
cc @Amruta-Ranade - we're planning to have this in the point release (19.1.1).

**Review by jordanlewis — COMMENTED** [review]
:lgtm_strong: 

The tests are failing because I think the results of the query to `node_statement_statistics` can return in arbitrary order - to fix that, just add a `rowsort` declaration to the test. So it'd be `query TB colnames rowsort`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36341#-:-Lc1zC84-uKk6vTnOeeS:bmod8tu)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @nvanbenschoten)

**Comment by celiala** [comment]
TFTR!

adding `rowsort` fixed the test, thx!

bors=r+

**Review by nvb — COMMENTED** [review]
Thanks for fixing this!

The one thing I don't see is where we add a new column on the `/statements` page and where we ensure that the same statement executed sometimes in implicit transactions and sometimes in explicit transactions gets split out into two different statement types. Does this PR do that?

Also, I think the issue this claims to fix is wrong.

> Reviewed 13 of 13 files at r2, 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36341#-:-Lc7G1PU1HPVTN0yxSSN:b-64azz1)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @celiala)

___
*[pkg/sql/crdb_internal.go, line 595 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36341#-Lc7FaTY3TJXFwsuOgbQ:-Lc7FaTY3TJXFwsuOgbR:b-t37jn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/eee20a805f9b638f2a40f474f16421f7af3dbe0b/pkg/sql/crdb_internal.go#L595)):*
> ```Go
>   overhead_lat_avg    FLOAT NOT NULL,
>   overhead_lat_var    FLOAT NOT NULL,
>   implicit_txn        BOOL
> ```

This should be `NOT NULL`, right?

**Comment by celiala** [comment]
TFTR!

> issue this claims to fix is wrong.

oops - fixed :)

> The one thing I don't see is where we add a new column on the /statements page 

ah, yes: we do need an `implicit txn` column on the statements page; will do.

> implicit_txn        BOOL

yes, should be `NOT NULL` - will fix.

**Comment by nvanbenschoten** [comment]
I just noticed that this PR also fixes https://github.com/cockroachdb/cockroach/issues/36286.

**Comment by celiala** [comment]
Hi all - this is RFAL, thanks!

* ui: Added a new column "Implicit Txn" on `/statements` page, to split these out into two different statement types
* fixed `implicit_txn` column on `crdb_internal.node_statement_statistics` to type `BOOL NOT NULL`

**Comment by piyush-singh** [comment]
A small nit: I think this might be more readable if the column is instead "Transaction Type" ("Txn Type" works if we want to conserve space) and the rows are marked as "Explicit" or "Implicit". This way, when users scroll they won't need to scroll up to see whether the question for the column is for explicit or implicit. 

@Amruta-Ranade we should probably add a tooltip for the column header as well. My rough attempt to write up a tooltip: `Statements in explicit transactions are wrapped by BEGIN and COMMIT statements by the client. CockroachDB wraps individual statements in implicit transactions. Explicit transactions employ [transactional pipelining](https://www.cockroachlabs.com/docs/v19.1/architecture/transaction-layer.html#transaction-pipelining) and therefore report latencies that do not account for replication.`. Feel free to suggest any edits.

File: pkg/ui/src/views/statements/statementsTable.tsx [hunk 2]
```diff
@@ -75,7 +77,13 @@ export function makeStatementsColumns(statements: AggregateStatistics[], selecte
...
       title: "Statement",
       className: "statements-table__col-query-text",
-      cell: (stmt) => <StatementLink statement={ stmt.label } app={ selectedApp } />,
+      cell: (stmt) => (
+        <StatementLink
+          statement={ stmt.label }
+          implicitTxn={ stmt.implicitTxn }
+          app={ selectedApp }
+        />
+        ),
```
Comment (vilterp) [inline pkg/ui/src/views/statements/statementsTable.tsx]: nit: outdent

File: pkg/ui/src/views/statements/statementsTable.tsx [hunk 3]
```diff
@@ -129,6 +137,12 @@ function makeCommonColumns(statements: AggregateStatistics[])
   const latencyBar = latencyBarChart(statements);
 
   return [
+    {
```
Comment (vilterp) [inline pkg/ui/src/views/statements/statementsTable.tsx]: I'd say while we're here, change the default sort column to latency (it's what people care about…)

File: pkg/ui/src/views/statements/statementDetails.tsx [hunk 4]
```diff
@@ -344,6 +347,10 @@ class StatementDetails extends React.Component<StatementDetailsProps, StatementD
                   { intersperse<ReactNode>(app.map(a => <AppLink app={ a } key={ a } />), ", ") }
                 </td>
               </tr>
+              <tr className="numeric-stats-table__row--body">
+                <th className="numeric-stats-table__cell" style={{ textAlign: "left" }}>Transaction Type</th>
```
Comment (vilterp) [inline pkg/ui/src/views/statements/statementDetails.tsx]: why does this column need to exist? won't it be the same for each row?

File: pkg/ui/src/views/statements/statementsPage.tsx [hunk 5]
```diff
@@ -59,7 +69,7 @@ class StatementsPage extends React.Component<StatementsPageProps & RouteProps, S
     super(props);
     this.state = {
       sortSetting: {
-        sortKey: 1,
+        sortKey: 2,
```
Comment (vilterp) [inline pkg/ui/src/views/statements/statementsPage.tsx]: Honestly, I think latency should be the default sort (it's what people care about; I don't think "total time" is that meaningful…)

**Comment by celiala** [comment]
TFTR!

Updates:
- Statement Summary/Details page: Latency is default sorted column
- Statement Details: removed Txn Type column
- Statement Summary: Txn Type column: addded Tooltip (see screenshot)
- fixed outdent nit

**Comment by Amruta-Ranade** [comment]
@piyush-singh @celiala  I need help understanding the context for the tooltip - can we sync on Monday to figure it out?

**Comment by celiala** [comment]
bors r+

@Amruta-Ranade and I synced offline: I'll merge feature now, so that this gets into the alpha cut. And we'll sync on context sometime this week, to adjust copy/documentation as needed for next cut.

**Pull request merged**